### DISCUSSION TITLE
Feature Request: Enhanced Filtering Capabilities for BaseModel Collections

### DISCUSSION THREAD

**Discussion by nh916** [pydantic/pydantic#9027] [Feature Request]
Dear Pydantic Contributors and Community,

I hope this message finds you well. I'm reaching out to propose a feature that could significantly enhance the utility and flexibility of Pydantic, particularly when working with collections of model instances.

In many scenarios, I've encountered the need to filter through lists of Pydantic models based on various criteria. While Pydantic provides exceptional tools for data validation and parsing, extending its functionality to include filtering directly within the BaseModel that all models would inherit from could be invaluable.

This feature request proposes the introduction of enhanced filtering capabilities to Pydantic's BaseModel to facilitate more flexible data handling and manipulation directly within the model's context. Specifically, the addition of generic filtering methods that support both attribute-based criteria and custom logic through callable functions would significantly benefit data processing tasks.

## Motivation

In many applications, processing lists of Pydantic model instances with specific filtering criteria is a common requirement. Currently, developers must implement external filtering logic, which can lead to repetitive code patterns and reduce overall code clarity and maintainability. By integrating advanced filtering methods directly into Pydantic's `BaseModel`, we can provide a more streamlined and efficient approach to handling model collections.

### Solution 1: Attribute and Callable-Based Filtering

This approach allows filtering based on direct attribute values or more complex conditions using callable functions.

```python
from typing import Any, List, TypeVar, Callable
from pydantic import BaseModel

T = TypeVar("T", bound='BaseModel')

class FilterableBaseModel(BaseModel):
    @classmethod
    def filter(cls, items: List[T], **criteria: Any) -> List[T]:
        """
        Filters a list of BaseModel-derived objects based on multiple criteria,
        supporting both direct value comparisons and callable functions for complex filtering logic.

        Args:
            items: The list of items to filter.
            **criteria: Attribute names and their expected values for filtering, or callable functions
                        that take an attribute value and return a boolean.

        Returns:
            A list of filtered items.

        Examples:
            # Filtering by attribute value
            >>> active_items = ExampleModel.filter(items, is_active=True)
            # Filtering by function
            >>> special_items = ExampleModel.filter(items, type=lambda x: x.startswith("HELLO"))
        """
        def matches(item: T, attribute: str, value: Any) -> bool:
            attr_value = getattr(item, attribute, None)
            return value(attr_value) if callable(value) else attr_value == value

        return [item for item in items if all(matches(item, attr, val) for attr, val in criteria.items())]
```

### Solution 2: Filtering with a Custom Function

This version emphasizes the use of a custom function to evaluate each item, providing maximum flexibility.

```python
from typing import List, TypeVar, Callable
from pydantic import BaseModel

T = TypeVar("T", bound='BaseModel')

class FilterableBaseModel(BaseModel):
    @classmethod
    def filter(cls, items: List[T], filter_fn: Callable[[T], bool]) -> List[T]:
        """
        Filters a list of BaseModel-derived objects based on a provided filtering function.

        Args:
            items: The list of items to filter.
            filter_fn: A callable that accepts an instance of T and returns True if the item should
                       be included in the result, otherwise False.

        Returns:
            A list of filtered items.

        Examples:
            # Define a custom filter function
            def active_with_type_prefix(item: ExampleModel) -> bool:
                return item.is_active and item.type.startswith("SOME_PREFIX")

            # Use the filter method
            filtered_items = ExampleModel.filter(items, active_with_type_prefix)
        """
        return [item for item in items if filter_fn(item)]
```

## Use Cases and Benefits

Both proposed solutions aim to enhance data processing flexibility within the Pydantic framework, accommodating simple attribute-based filtering as well as complex, condition-based logic. This feature would be particularly useful for applications dealing with large sets of model instances requiring dynamic filtering, such as web applications, data analysis tools, and APIs.

## Request for Feedback

I am keen to hear the community's thoughts on these proposals, including preferences between the two solutions, or any ideas/suggestions.

**Comment by wishbonesr** [comment]
I'm surprised there isn't more feedback, as I immediately needed filtering as there are inevitable API calls that do not provide parameterization levels necessary for every situation.  Additionally we could be consuming data from other sources (e.g. DB) in which development is frozen and filtering criteria there as well will be a manual process.

I support the  Solution 1 proposed.

**Comment by Viicos** [comment]
Looks like this can easily be implemented as a standalone third-party library?
